### PR TITLE
fix(redirect-plugin): redirection loop behind a proxy or lb

### LINKED ISSUE #6078: request help: Use plugin redirect for http_to_https,the browser access unlimited 301.
### Issue description

Use plugin redirect for http_to_https,the browser access unlimited 301.

my apisix's config,e.g.:
`{
    "create_time": 1641884664,
    "desc": "",
    "upstream_id": "00000000000001329854",
    "enable_websocket": false,
    "plugin_config_id": "1688",
    "plugins": {
        "redirect": {
            "ret_code": 301,
            "http_to_https": true,
            "encode_uri": false,
            "append_query_string": false
        },
        "proxy-rewrite": {
            "headers": {
                "primeval_host": "$host"
            },
            "disable": false
        }
    },
    "priority": 0,
    "update_time": 1641893364,
    "id": "00000000000000285664",
    "name": "mydashboard",
    "uris": [
        "/mydashboard/*"
    ],
    "status": 1,
    "methods": [
        "GET",
        "HEAD",
        "POST",
        "PUT",
        "DELETE",
        "OPTIONS",
        "PATCH"
    ],
    "hosts": [
        "ttt3.mydashboard.com"
    ]
}`

### Environment

- apisix version (cmd: `apisix version`): apisix2.10.1
- OS (cmd: `uname -a`): centos 7.3
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

### PR DESCRIPTION
**Pull request by soham4abc**
### What this PR does / why we need it:
Fixes #6078 

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: docs/en/latest/architecture-design/apisix.md [hunk 1]
```diff
@@ -25,6 +25,10 @@ title: APISIX
 
 ![flow-software-architecture](../../../assets/images/flow-software-architecture.png)
 
+## Apache APISIX : Software Architecture
+
+![flow-software-architecture](../../../assets/images/flow-software-architecture.png)
+
```
Comment (bisakhmondal) [inline docs/en/latest/architecture-design/apisix.md:31]: Redundant?

**Review by bisakhmondal — COMMENTED** [review]
Thanks for the PR

File: docs/en/latest/architecture-design/apisix.md (hunk 1 above)
Comment (soham4abc) [reply docs/en/latest/architecture-design/apisix.md:31]: Yes, I am removing this in a minute

**Review by bisakhmondal — DISMISSED** [review]
lgtm

File: docs/en/latest/architecture-design/apisix.md (hunk 1 above)
Comment (bisakhmondal) [reply docs/en/latest/architecture-design/apisix.md]: Thanks for addressing the change

File: apisix/plugins/redirect.lua [hunk 2]
```diff
@@ -150,7 +150,9 @@ function _M.rewrite(conf, ctx)
     local uri = conf.uri
     local regex_uri = conf.regex_uri
 
-    if conf.http_to_https and ctx.var.scheme == "http" then
+    local proxy_proto = core.request.header(ctx, "x-forwarded-proto")
+    local _scheme = proxy_proto and proxy_proto or ctx.var.scheme
```
Comment (spacewander) [inline apisix/plugins/redirect.lua:154]:
```suggestion
    local _scheme = proxy_proto or ctx.var.scheme
```

**Review by spacewander — COMMENTED** [review]
@bisakhmondal 
Could you help @soham4abc to add a test for it?

**Comment by bisakhmondal** [comment]
> @bisakhmondal Could you help @soham4abc to add a test for it?

Yes sure. @soham4abc we use test-nginx in APISIX for testing. You can find the docs here (https://github.com/apache/apisix/blob/master/docs/en/latest/internal/testing-framework.md)

Feel free to go through the existing redirect plugin tests here (https://github.com/apache/apisix/blob/master/t/plugin/redirect.t) and add a test case there. I can think of two options:
1. Treating the test suite endpoint itself as a proxy
2. Setup a mock HTTP route, that acts as a proxy to forward the request to the apisix test suite.

Totally up to you. Thank you!

**Comment by soham4abc** [comment]
> > @bisakhmondal Could you help @soham4abc to add a test for it?
> 
> Yes sure. @soham4abc we use test-nginx in APISIX for testing. You can find the docs here (https://github.com/apache/apisix/blob/master/docs/en/latest/internal/testing-framework.md)
> 
> Feel free to go through the existing redirect plugin tests here (https://github.com/apache/apisix/blob/master/t/plugin/redirect.t) and add a test case there. I can think of two options:
> 
> 1. Treating the test suite endpoint itself as a proxy
> 2. Setup a mock HTTP route, that acts as a proxy to forward the request to the apisix test suite.
> 
> Totally up to you. Thank you!

Hey @bisakhmondal ,

Sorry as I am new to this codebase and do not have much experience regarding it. Would be nice if you raise a separate issue and PR for this. 

I am sorry I couldn't deliver the testcases as of now!

Thank you

**Comment by soham4abc** [comment]
Any issues? @bisakhmondal

**Comment by spacewander** [comment]
Adding a test for it is easy. Could you follow @bisakhmondal 's guide to add a test?
A test is required to prevent regression.

**Comment by soham4abc** [comment]
> Adding a test for it is easy. Could you follow @bisakhmondal 's guide to add a test?
> A test is required to prevent regression.

Sure... Just might take some time!

File: apisix/plugins/redirect.lua [hunk 3]
```diff
@@ -150,7 +150,9 @@ function _M.rewrite(conf, ctx)
     local uri = conf.uri
     local regex_uri = conf.regex_uri
 
-    if conf.http_to_https and ctx.var.scheme == "http" then
+    local proxy_proto = core.request.header(ctx, "x-forwarded-proto")
+    local _scheme = proxy_proto or ctx.var.scheme
```
Comment (tzssangglass) [inline apisix/plugins/redirect.lua:154]:
```suggestion
    local proxy_proto = core.request.header(ctx, "X-Forwarded-Proto")
    local _scheme = proxy_proto or core.request.get_scheme(ctx)
```

File: t/plugin/redirect.t [hunk 4]
```diff
@@ -1000,3 +1000,44 @@ Location: /hello?type=string&name=json
...
 --- no_error_log
 [error]
+
+
+
+=== TEST 41: redirect
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("plugin.redirect").test
```
Comment (tzssangglass) [inline t/plugin/redirect.t:1010]:
```suggestion
            local t = require("lib.test_admin").test
```

File: t/plugin/redirect.t [hunk 5]
```diff
@@ -1000,3 +1000,44 @@ Location: /hello?type=string&name=json
...
+                        [
+                            "scheme",
+                            "==",
+                            "http"
+                        ]
+                    ],
+                    "plugins": {
+                        "redirect": {
+                            "uri": "https://$host$request_uri",
+                            "ret_code": 301
```
Comment (tzssangglass) [inline t/plugin/redirect.t:1026]:
```suggestion
                            "http_to_https": true
```

File: t/plugin/redirect.t [hunk 6]
```diff
@@ -1000,3 +1000,44 @@ Location: /hello?type=string&name=json
...
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
```
Comment (tzssangglass) [inline t/plugin/redirect.t:1043]: and add a test case like 


```
=== TEST 2: enable http_to_https (pass X-Forwarded-Proto)
--- request
GET /hello
--- more_headers
Host: foo.com
X-Forwarded-Proto: http
--- error_code: 301
--- response_headers
Location: https://foo.com/hello
```

File: t/plugin/redirect.t [hunk 7]
```diff
@@ -1000,3 +1000,44 @@ Location: /hello?type=string&name=json
 --- error_code: 302
 --- no_error_log
 [error]
+
+
+
+=== TEST 41: redirect
```
Comment (tzssangglass) [inline t/plugin/redirect.t:1006]:
```suggestion
=== TEST 41: enable http_to_https (pass X-Forwarded-Proto)
```

File: t/plugin/redirect.t [hunk 8]
```diff
@@ -1000,3 +1000,49 @@ Location: /hello?type=string&name=json
...
+                }]]
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+
```
Comment (tzssangglass) [inline t/plugin/redirect.t:1037]: lost 

```
--- request
GET /t
--- response_body
passed
--- no_error_log
[error]
```

**Review by tzssangglass — DISMISSED** [review]

**Review by starsz — APPROVED** [review]

**Comment by soham4abc** [comment]
Thanks a lot @tzssangglass  and @bisakhmondal  For guiding me in this issue!

File: t/plugin/redirect.t [hunk 9]
```diff
@@ -1000,3 +1000,58 @@ Location: /hello?type=string&name=json
...
```
Comment (spacewander) [inline t/plugin/redirect.t:1037]: Let's remove the extra blank lines

Comment (soham4abc) [reply t/plugin/redirect.t:1037]: @spacewander  Won't removing the extra lines make lint errors??

Comment (soham4abc) [reply t/plugin/redirect.t:1037]: You meant line 1030 right??

Comment (spacewander) [reply t/plugin/redirect.t:1037]: I mean L1037 to L1039

Comment (soham4abc) [reply t/plugin/redirect.t]: Done as per request

**Review by spacewander — APPROVED** [review]

**Review by bisakhmondal — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Comment by soham4abc** [comment]
Thank you everyone

**Pull request merged**